firedrake/mesh.py

```diff
@@ -594,10 +620,135 @@ def create_cell_node_list(self, global_numbering, fiat_element):
+        .. note::
+
+            This only returns the dimension of the "side" (vertical) facets,
+            not the "top" or "bottom" (horizontal) facets.
+
+        """
+        return (self._base_mesh.facet_dimension(), 1)
+
+    @utils.cached_property
+    def cell_set(self):
+        return op2.ExtrudedSet(self._base_mesh.cell_set, layers=self.layers)
 
     def _order_data_by_cell_index(self, column_list, cell_data):
-        return cell_data[column_list]
+        cell_list = []
+        for col in column_list:
+            cell_list += range(col, col + (self.layers - 1))
+        return cell_data[cell_list]
+
+
+class MeshGeometry(object):
+    """A representation of mesh topology and geometry."""
+
+    def __new__(cls, *args, **kwargs):
+        """Create mesh geometry object."""
+        utils._init()
+        mesh = super(MeshGeometry, cls).__new__(cls)
+        mesh.uid = utils._new_uid()
+        return mesh
+
+    def __init__(self, coordinates):
+        """Initialise a mesh geometry from coordinates.
+
+        :arg coordinates: a coordinateless function containing the coordinates
+        """
+        # Direct link to topology
+        self._topology = coordinates.function_space().mesh()
+
+        # Cache mesh object on the coordinateless coordinates function
+        coordinates._as_coordinates = weakref.ref(self)
+
+        self._coordinates = coordinates
+        self._ufl_domain = ufl.Domain(coordinates)
+
+        # Add subdomain_data to the measure objects we store with
+        # the mesh.  These are weakrefs for consistency with the
+        # "global" measure objects
+        self._dx = ufl.Measure('cell', subdomain_data=weakref.ref(coordinates))
+        self._ds = ufl.Measure('exterior_facet', subdomain_data=weakref.ref(coordinates))
+        self._dS = ufl.Measure('interior_facet', subdomain_data=weakref.ref(coordinates))
+        if self.layers:
+            self._ds_t = ufl.Measure('exterior_facet_top', subdomain_data=weakref.ref(coordinates))
+            self._ds_b = ufl.Measure('exterior_facet_bottom', subdomain_data=weakref.ref(coordinates))
+            self._ds_v = ufl.Measure('exterior_facet_vert', subdomain_data=weakref.ref(coordinates))
+            self._dS_h = ufl.Measure('interior_facet_horiz', subdomain_data=weakref.ref(coordinates))
+            self._dS_v = ufl.Measure('interior_facet_vert', subdomain_data=weakref.ref(coordinates))
+        # Set the subdomain_data on all the default measures to this
+        # coordinate field.
+        # We don't set the domain on the measure since this causes
+        # an uncollectable reference in the global space (dx is
+        # global).  Furthermore, it's never used anyway.
+        for measure in [ufl.dx, ufl.ds, ufl.dS]:
+            measure._subdomain_data = weakref.ref(coordinates)
+        if self.layers:
+            for measure in [ufl.ds_t, ufl.ds_b, ufl.ds_v, ufl.dS_h, ufl.dS_v]:
+                measure._subdomain_data = weakref.ref(coordinates)
```

Comment:
I /think/ that since we no longer get coordinates from domain().data().coordinates, we don't need to pass subdomain_data in at all?